File: Problems/0100. (E) Same Tree/Java/solution.java
class Solution {
  public boolean isSameTree(TreeNode p, TreeNode q) {
    // recursively compare nodes

    // base case, both nodes are null
    if (p == null && q == null)
      return true;
    // base case, only one node is null
    if (p == null || q == null)
      return false;
    // nodes match, recurse on subtrees
    if (p.val == q.val)
      return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    // base case, nodes do not match
    return false;
  }
}

File: Problems/0100. (E) Same Tree/Python3/solution.py
class Solution:
  def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    # stacks

    stack_p = [p]
    stack_q = [q]
    while stack_p:
      cur_p = stack_p.pop()
      cur_q = stack_q.pop()
      if cur_p is None and cur_q is None:
        continue
      elif cur_p is None or cur_q is None:
        return False
      else:
        if cur_p.val != cur_q.val:
          return False
        stack_p.append(cur_p.right)
        stack_p.append(cur_p.left)
        stack_q.append(cur_q.right)
        stack_q.append(cur_q.left)

    return True

